test_asyncio fails

======================================================================
FAIL: test_log_destroyed_pending_task (test.test_asyncio.test_tasks.TaskTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\brcan\Documents\Repositories\Pyjion\Python\lib\test\test_asyncio\test_tasks.py", line 1693, in test_log
_destroyed_pending_task
    self.assertEqual(asyncio.Task.all_tasks(loop=self.loop), set())
AssertionError: Items in the first set but not the second:
<Task pending coro=<kill_me() running at C:\Users\brcan\Documents\Repositories\Pyjion\Python\lib\test\test_asyncio\test_
tasks.py:1666> wait_for=<Future pending cb=[Task._wakeup()] created at C:\Users\brcan\Documents\Repositories\Pyjion\Pyth
on\lib\test\test_asyncio\test_tasks.py:1665> created at C:\Users\brcan\Documents\Repositories\Pyjion\Python\lib\test\tes
t_asyncio\test_tasks.py:1677>

======================================================================
FAIL: test_task_repr (test.test_asyncio.test_tasks.TaskTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\brcan\Documents\Repositories\Pyjion\Python\lib\test\test_asyncio\test_tasks.py", line 192, in test_task
_repr
    coro_qualname)
AssertionError: 'notmuch' != 'TaskTests.test_task_repr.<locals>.notmuch'
- notmuch
+ TaskTests.test_task_repr.<locals>.notmuch


======================================================================
FAIL: test_task_repr_coro_decorator (test.test_asyncio.test_tasks.TaskTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\brcan\Documents\Repositories\Pyjion\Python\lib\test\test_asyncio\test_tasks.py", line 254, in test_task
_repr_coro_decorator
    self.assertEqual(gen.__name__, coro_name)
AssertionError: 'coro' != 'notmuch'
- coro
+ notmuch


----------------------------------------------------------------------

--- a/Pyjion/absint.cpp
+++ b/Pyjion/absint.cpp
@@ -1540,8 +1540,41 @@ void AbstractInterpreter::build_list(size_t argCnt) {
 void AbstractInterpreter::build_set(size_t argCnt) {
     m_comp->emit_new_set();
     error_check("build set failed");
-    m_comp->emit_set_store(argCnt);
-    dec_stack(argCnt);
+
+    if (argCnt != 0) {
+        auto valueTmp = m_comp->emit_define_local();
+        auto setTmp = m_comp->emit_define_local();
+
+        m_comp->emit_store_local(setTmp);
+
+        for (size_t i = 0, arg = argCnt - 1; i < argCnt; i++, arg--) {
+            // save the argument into a temporary...
+            m_comp->emit_store_local(valueTmp);
+
+            dec_stack();
+
+            // load the address of the tuple item...
+            m_comp->emit_load_local(setTmp);
+            m_comp->emit_load_local(valueTmp);
+            m_comp->emit_set_add();
+
+            auto noErr = m_comp->emit_define_label();
+            m_comp->emit_branch(BranchTrue, noErr);
+            
+            // free the set too
+            m_comp->emit_load_local(setTmp);
+            m_comp->emit_pop_top();
+
+            branch_raise("set add failed");
+            m_comp->emit_mark_label(noErr);
+
+        }
+
+        m_comp->emit_load_local(setTmp);
+        m_comp->emit_free_local(valueTmp);
+        m_comp->emit_free_local(setTmp);
+    }
+    inc_stack();
 }
 
 
@@ -2018,7 +2051,7 @@ JittedCode* AbstractInterpreter::compile_worker() {
                 m_comp->emit_build_slice();
                 inc_stack();
                 break;
-            case BUILD_SET: build_set(oparg); inc_stack(); break;
+            case BUILD_SET: build_set(oparg); break;
             case UNARY_POSITIVE: unary_positive(opcodeIndex); break;
             case UNARY_NEGATIVE: unary_negative(opcodeIndex); break;
             case UNARY_NOT: unary_not(curByte); break;

--- a/Pyjion/ipycomp.h
+++ b/Pyjion/ipycomp.h
@@ -219,8 +219,6 @@ public:
 
     // Creates a new set
     virtual void emit_new_set() = 0;
-    // Stores all of the values on the stack into a set
-    virtual void emit_set_store(size_t argCnt) = 0;
     // Adds a single item to a set
     virtual void emit_set_add() = 0;
 

--- a/Pyjion/pycomp.cpp
+++ b/Pyjion/pycomp.cpp
@@ -375,30 +375,6 @@ void PythonCompiler::emit_new_set() {
     m_il.emit_call(METHOD_PYSET_NEW);
 }
 
-void PythonCompiler::emit_set_store(size_t argCnt) {
-    if (argCnt != 0) {
-        auto valueTmp = m_il.define_local(Parameter(CORINFO_TYPE_NATIVEINT));
-        auto setTmp = m_il.define_local(Parameter(CORINFO_TYPE_NATIVEINT));
-
-        m_il.st_loc(setTmp);
-
-        for (size_t i = 0, arg = argCnt - 1; i < argCnt; i++, arg--) {
-            // save the argument into a temporary...
-            m_il.st_loc(valueTmp);
-
-            // load the address of the tuple item...
-            m_il.ld_loc(setTmp);
-            m_il.ld_loc(valueTmp);
-            m_il.emit_call(METHOD_PYSET_ADD);
-            m_il.pop();
-        }
-
-        m_il.ld_loc(setTmp);
-        m_il.free_local(valueTmp);
-        m_il.free_local(setTmp);
-    }
-}
-
 void PythonCompiler::emit_new_dict(size_t size) {
     m_il.ld_i(size);
     m_il.emit_call(METHOD_PYDICT_NEWPRESIZED);
@@ -840,8 +816,7 @@ void PythonCompiler::emit_set_add() {
     m_il.emit_call(METHOD_SET_ADD_TOKEN);
 }
 
-void PythonCompiler::emit_map_add()
-{
+void PythonCompiler::emit_map_add() {
     m_il.emit_call(METHOD_MAP_ADD_TOKEN);
 }
 

--- a/Pyjion/pycomp.h
+++ b/Pyjion/pycomp.h
@@ -267,7 +267,6 @@ public:
     virtual void emit_list_store(size_t argCnt);
 
     virtual void emit_new_set();
-    virtual void emit_set_store(size_t argCnt);
     virtual void emit_dict_store();
 
     virtual void emit_new_dict(size_t size);

